stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOM.java

```diff
@@ -208,6 +214,96 @@ public void run() {
     });
   }
 
+  @ChromeDevtoolsMethod
+  public PerformSearchResponse performSearch(JsonRpcPeer peer, JSONObject params) {
+    final PerformSearchRequest request = mObjectMapper.convertValue(
+        params,
+        PerformSearchRequest.class);
+
+    // Searches are case insensitive so convert
+    // the query to lower case for simpler comparisons.
+    request.query = request.query.toLowerCase();
+
+    final Object root = this.mShadowDOM.getRootElement();
+
+    // Each search action has a unique ID so that
+    // it can be queried later.
+    final String id = UUID.randomUUID().toString();
+
+    final ArrayList<Integer> results = new ArrayList<>();
+    FindMatches(request.query, root, results);
+
+    mSearchResults.put(id, results);
+
+    final PerformSearchResponse response = new PerformSearchResponse();
+    response.searchId = id;
+    response.resultCount = results.size();
+
+    return response;
+  }
+
+  private void FindMatches(String query, Object root, ArrayList<Integer> resultIds) {
+    final ElementInfo info = mShadowDOM.getElementInfo(root);
+
+    for (int i = 0; i < info.children.size(); i++) {
+      final Object element = info.children.get(i);
+
+      if (DoesElementMatch(query, element)) {
+        resultIds.add(mObjectIdMapper.getIdForObject(element));
+      }
+
+      FindMatches(query, element, resultIds);
+    }
+  }
+
+  private boolean DoesElementMatch(String query, Object element) {
+    final String queryParam = query;
+    final Object elemParam = element;
+
+    return mDOMProvider.postAndWait(new UncheckedCallable<Boolean>() {
```

Comment:
I recommend putting the entire search into a `postAndWait` callback, as there's some decent overhead to this thing. To do this, remove the post-and-waiting from `doesElementMatch` and hoist it into the calling method, `findMatches` (that whole method should be inside a post-and-wait callback, from the looks of it). That way you'll have 1 post-and-wait for the whole search instead of 1 per element.